### PR TITLE
Support for PHP 7.1

### PR DESCRIPTION
**Pull request by prolic**
- add support for PHP 7.1
- uses custom prophecy branch (see https://github.com/phpspec/prophecy/pull/287)
- uses custom prooph/common branch ("prooph/common" : "dev-php_7_1 as 4.0")
- create ProophActionEventEmitter with available event names (see https://github.com/prooph/service-bus/issues/105)

### CODE REVIEW

**Comment by prolic** [comment]
tests passing with PHP 7.1.0RC3, but unfortunately travis currently uses PHP 7.1.0RC2

File: src/Plugin/Router/RegexRouter.php [hunk 1]
```diff
@@ -46,37 +46,29 @@ class RegexRouter implements MessageBusRouterPlugin, ActionEventListenerAggregat
...
 
-    /**
-     * @param string $pattern
-     * @return $this
-     * @throws Exception\RuntimeException
-     */
-    public function route($pattern)
+    public function route(string $pattern) : RegexRouter
     {
         Assertion::string($pattern);
```
Comment (oqq) [inline src/Plugin/Router/RegexRouter.php]: This call will be unnecessary with type hinted argument.

File: src/Plugin/Router/SingleHandlerRouter.php [hunk 2]
```diff
@@ -41,31 +41,23 @@ class SingleHandlerRouter implements MessageBusRouterPlugin, ActionEventListener
...
 
-    /**
-     * @param string $messageName
-     * @return $this
-     * @throws Exception\RuntimeException
-     */
-    public function route($messageName)
+    public function route(string $messageName) : SingleHandlerRouter
     {
         Assertion::string($messageName);
```
Comment (oqq) [inline src/Plugin/Router/SingleHandlerRouter.php]: This call will be unnecessary with type hinted argument.

Comment (prolic) [inline src/Plugin/Router/SingleHandlerRouter.php]: fixed

File: src/MessageBus.php [hunk 3]
```diff
@@ -49,31 +51,27 @@
...
-     */
-    public function deactivate(ActionEventListenerAggregate $plugin)
+    public function deactivate(ActionEventListenerAggregate $plugin): void
     {
         $plugin->detach($this->getActionEventEmitter());
     }
 
     /**
      * @param mixed $message
+     *
```
Comment (oqq) [inline src/MessageBus.php]: empty line between equal annotation tag types

File: src/Plugin/Guard/AuthorizationService.php [hunk 4]
```diff
@@ -20,8 +20,10 @@
      * Check if the permission is granted to the current identity
      *
      * @param string $messageName
+     *
```
Comment (oqq) [inline src/Plugin/Guard/AuthorizationService.php]: empty line between equal annotation tag types

File: src/Plugin/InvokeStrategy/AbstractInvokeStrategy.php [hunk 5]
```diff
@@ -33,33 +33,28 @@
 
     /**
      * @param mixed $handler
+     *
```
Comment (oqq) [inline src/Plugin/InvokeStrategy/AbstractInvokeStrategy.php]: empty line between equal annotation tag types

File: src/Plugin/InvokeStrategy/HandleCommandStrategy.php [hunk 6]
```diff
@@ -22,10 +22,12 @@ class HandleCommandStrategy extends AbstractInvokeStrategy
 {
     /**
      * @param mixed $handler
+     *
```
Comment (oqq) [inline src/Plugin/InvokeStrategy/HandleCommandStrategy.php]: empty line between equal annotation tag types

File: src/Plugin/InvokeStrategy/HandleCommandStrategy.php [hunk 7]
```diff
@@ -34,9 +36,12 @@ public function canInvoke($handler, $message)
 
     /**
      * @param mixed $handler
+     *
```
Comment (oqq) [inline src/Plugin/InvokeStrategy/HandleCommandStrategy.php]: empty line between equal annotation tag types

File: src/Plugin/InvokeStrategy/OnEventStrategy.php [hunk 8]
```diff
@@ -22,10 +22,12 @@ class OnEventStrategy extends AbstractInvokeStrategy
 {
     /**
      * @param mixed $handler
+     *
```
Comment (oqq) [inline src/Plugin/InvokeStrategy/OnEventStrategy.php]: empty line between equal annotation tag types

File: src/Plugin/Router/EventRouter.php [hunk 9]
```diff
@@ -94,11 +85,14 @@ public function route($eventName)
 
     /**
      * @param string|object|callable $eventListener
-     * @return $this
+     *
+     * @return EventRouter
+     *
      * @throws Exception\RuntimeException
+     *
```
Comment (oqq) [inline src/Plugin/Router/EventRouter.php]: empty line between equal annotation tag types

File: src/Plugin/Router/RegexRouter.php [hunk 10]
```diff
@@ -92,11 +83,14 @@ public function route($pattern)
 
     /**
      * @param string|object|callable $handler
-     * @return $this
+     *
+     * @return RegexRouter
+     *
      * @throws Exception\RuntimeException
+     *
```
Comment (oqq) [inline src/Plugin/Router/RegexRouter.php]: empty line between equal annotation tag types

File: src/Plugin/Router/SingleHandlerRouter.php [hunk 11]
```diff
@@ -81,11 +72,14 @@ public function route($messageName)
 
     /**
      * @param string|object|callable $messageHandler
-     * @return $this
+     *
+     * @return SingleHandlerRouter
+     *
      * @throws Exception\RuntimeException
+     *
```
Comment (oqq) [inline src/Plugin/Router/SingleHandlerRouter.php]: empty line between equal annotation tag types

File: src/Plugin/Router/SingleHandlerRouter.php [hunk 12]
```diff
@@ -97,7 +91,9 @@ public function to($messageHandler)
         if (null === $this->tmpMessageName) {
             throw new Exception\RuntimeException(sprintf(
                 "Cannot map handler %s to a message. Please use method route before calling method to",
-                (is_object($messageHandler))? get_class($messageHandler) : (is_string($messageHandler))? $messageHandler : gettype($messageHandler)
+                (is_object($messageHandler))
```
Comment (oqq) [inline src/Plugin/Router/SingleHandlerRouter.php]: unnecessary brackets

File: src/Plugin/Router/SingleHandlerRouter.php [hunk 13]
```diff
@@ -97,7 +91,9 @@ public function to($messageHandler)
...
+                    ? get_class($messageHandler)
+                    : (is_string($messageHandler))? $messageHandler : gettype($messageHandler)
```
Comment (oqq) [inline src/Plugin/Router/SingleHandlerRouter.php]: unnecessary brackets around "is_string"

File: src/MessageBus.php (hunk 3 above)
Comment (prolic) [inline src/MessageBus.php]: fixed

File: src/Plugin/Guard/AuthorizationService.php (hunk 4 above)
Comment (prolic) [inline src/Plugin/Guard/AuthorizationService.php]: fixed

File: src/Plugin/InvokeStrategy/HandleCommandStrategy.php (hunk 6 above)
Comment (prolic) [inline src/Plugin/InvokeStrategy/HandleCommandStrategy.php]: fixed

File: src/Plugin/InvokeStrategy/AbstractInvokeStrategy.php (hunk 5 above)
Comment (prolic) [inline src/Plugin/InvokeStrategy/AbstractInvokeStrategy.php]: fixed

File: src/Plugin/InvokeStrategy/OnEventStrategy.php (hunk 8 above)
Comment (prolic) [inline src/Plugin/InvokeStrategy/OnEventStrategy.php]: fixed

File: src/Plugin/Router/SingleHandlerRouter.php (hunk 12 above)
Comment (prolic) [inline src/Plugin/Router/SingleHandlerRouter.php]: fixed

File: src/Plugin/Router/SingleHandlerRouter.php (hunk 13 above)
Comment (prolic) [inline src/Plugin/Router/SingleHandlerRouter.php]: fixed

File: src/Plugin/Router/RegexRouter.php (hunk 10 above)
Comment (prolic) [inline src/Plugin/Router/RegexRouter.php]: fixed

File: src/Async/MessageProducer.php [hunk 14]
```diff
@@ -34,11 +35,6 @@
      *       MUST either be resolved/rejected OR the message producer
      *       MUST throw a Prooph\ServiceBus\Exception\RuntimeException if it cannot
      *       handle the $deferred
-     *
-     * @param Message $message
-     * @param null|Deferred $deferred
-     * @throws RuntimeException If a $deferred is passed but producer can not handle it
```
Comment (basz) [inline src/Async/MessageProducer.php]: why remove throws?

File: src/QueryBus.php [hunk 15]
```diff
@@ -105,7 +103,7 @@ public function dispatch($query)
             }
 
             $this->triggerFinalize($actionEvent);
-        } catch (\Exception $ex) {
+        } catch (\Throwable $ex) {
```
Comment (basz) [inline src/QueryBus.php]: import via use statement

File: src/QueryBus.php [hunk 16]
```diff
@@ -59,9 +55,11 @@ public function setActionEventEmitter(ActionEventEmitter $actionEventDispatcher)
 
     /**
      * @param mixed $query
+     *
      * @return Promise
```
Comment (basz) [inline src/QueryBus.php]: remove return

File: src/Plugin/Router/SingleHandlerRouter.php [hunk 17]
```diff
@@ -81,11 +72,13 @@ public function route($messageName)
 
     /**
      * @param string|object|callable $messageHandler
-     * @return $this
+     *
+     * @return SingleHandlerRouter
```
Comment (basz) [inline src/Plugin/Router/SingleHandlerRouter.php]: remove return

File: src/Plugin/Router/RegexRouter.php [hunk 18]
```diff
@@ -48,37 +48,28 @@ class RegexRouter implements MessageBusRouterPlugin, ActionEventListenerAggregat
...
-    public function attach(ActionEventEmitter $events)
+    public function attach(ActionEventEmitter $events): void
     {
         $this->trackHandler($events->attachListener(MessageBus::EVENT_ROUTE, [$this, 'onRouteMessage'], 100));
     }
 
-    /**
-     * @param string $pattern
-     * @return $this
-     * @throws Exception\RuntimeException
```
Comment (basz) [inline src/Plugin/Router/RegexRouter.php]: keep throws

File: src/Plugin/Router/EventRouter.php [hunk 19]
```diff
@@ -121,26 +116,25 @@ public function to($eventListener)
 
     /**
      * Alias for method to
+     *
      * @param string|object|callable $eventListener
-     * @return $this
+     *
+     * @return EventRouter
```
Comment (basz) [inline src/Plugin/Router/EventRouter.php]: remove return

File: src/Plugin/Router/EventRouter.php [hunk 20]
```diff
@@ -94,11 +85,13 @@ public function route($eventName)
 
     /**
      * @param string|object|callable $eventListener
-     * @return $this
+     *
+     * @return EventRouter
```
Comment (basz) [inline src/Plugin/Router/EventRouter.php]: remove return

File: src/Plugin/Router/EventRouter.php [hunk 21]
```diff
@@ -44,39 +44,30 @@ class EventRouter implements MessageBusRouterPlugin, ActionEventListenerAggregat
...
-    public function attach(ActionEventEmitter $events)
+    public function attach(ActionEventEmitter $events): void
     {
         $this->trackHandler($events->attachListener(MessageBus::EVENT_ROUTE, [$this, "onRouteMessage"]));
     }
 
-    /**
-     * @param string $eventName
-     * @return $this
-     * @throws Exception\RuntimeException
```
Comment (basz) [inline src/Plugin/Router/EventRouter.php]: keep throws

File: src/MessageBus.php [hunk 22]
```diff
@@ -130,59 +118,48 @@ protected function trigger(ActionEvent $actionEvent)
...
-    public function setActionEventEmitter(ActionEventEmitter $actionEventDispatcher)
+    public function setActionEventEmitter(ActionEventEmitter $actionEventDispatcher): void
     {
         $this->events = $actionEventDispatcher;
     }
 
-    /**
-     * Retrieve the action event dispatcher
-     *
-     * Lazy-loads a dispatcher if none is registered.
```
Comment (basz) [inline src/MessageBus.php]: might be useful to keep explanation, dunno

File: src/Plugin/Guard/AuthorizationService.php [hunk 23]
```diff
@@ -21,7 +21,8 @@
      *
      * @param string $messageName
      * @param mixed  $context
+     *
      * @return bool
```
Comment (basz) [inline src/Plugin/Guard/AuthorizationService.php]: remove param, param and return

File: src/Plugin/Guard/FinalizeGuard.php [hunk 24]
```diff
@@ -36,28 +36,20 @@
...
-    public function __construct(AuthorizationService $authorizationService, $exposeEventMessageName = false)
+    public function __construct(AuthorizationService $authorizationService, bool $exposeEventMessageName = false)
     {
         $this->authorizationService = $authorizationService;
         $this->exposeEventMessageName = $exposeEventMessageName;
     }
 
-    /**
-     * @param ActionEvent $actionEvent
-     * @throws UnauthorizedException
```
Comment (basz) [inline src/Plugin/Guard/FinalizeGuard.php]: keep throws

File: src/Plugin/InvokeStrategy/AbstractInvokeStrategy.php [hunk 25]
```diff
@@ -34,32 +34,25 @@
     /**
      * @param mixed $handler
      * @param mixed $message
+     *
      * @return bool
```
Comment (basz) [inline src/Plugin/InvokeStrategy/AbstractInvokeStrategy.php]: remove return

File: src/Plugin/InvokeStrategy/AbstractInvokeStrategy.php [hunk 26]
```diff
@@ -34,32 +34,25 @@
...
      * @param mixed $message
-     */
-    abstract protected function invoke($handler, $message);
-
-    /**
-     * Attach one or more listeners
-     *
-     * @param ActionEventEmitter $events
      *
      * @return void
```
Comment (basz) [inline src/Plugin/InvokeStrategy/AbstractInvokeStrategy.php]: remove void

File: src/Plugin/InvokeStrategy/FinderInvokeStrategy.php [hunk 27]
```diff
@@ -64,11 +58,15 @@ public function __invoke(ActionEvent $actionEvent)
 
     /**
      * @param mixed $query
+     *
      * @return string
```
Comment (basz) [inline src/Plugin/InvokeStrategy/FinderInvokeStrategy.php]: remove return

File: src/Plugin/InvokeStrategy/HandleCommandStrategy.php [hunk 28]
```diff
@@ -23,9 +23,10 @@ class HandleCommandStrategy extends AbstractInvokeStrategy
     /**
      * @param mixed $handler
      * @param mixed $message
+     *
      * @return bool
```
Comment (basz) [inline src/Plugin/InvokeStrategy/HandleCommandStrategy.php]: remove return

File: src/Plugin/InvokeStrategy/HandleCommandStrategy.php [hunk 29]
```diff
@@ -35,8 +36,10 @@ public function canInvoke($handler, $message)
     /**
      * @param mixed $handler
      * @param mixed $message
+     *
+     * @return void
```
Comment (basz) [inline src/Plugin/InvokeStrategy/HandleCommandStrategy.php]: remove void

File: src/Plugin/InvokeStrategy/HandleCommandStrategy.php [hunk 30]
```diff
@@ -49,11 +52,15 @@ public function invoke($handler, $message)
 
     /**
      * @param mixed $message
+     *
      * @return string
```
Comment (basz) [inline src/Plugin/InvokeStrategy/HandleCommandStrategy.php]: remove return

File: src/Plugin/InvokeStrategy/OnEventStrategy.php [hunk 31]
```diff
@@ -23,9 +23,10 @@ class OnEventStrategy extends AbstractInvokeStrategy
     /**
      * @param mixed $handler
      * @param mixed $message
+     *
      * @return bool
```
Comment (basz) [inline src/Plugin/InvokeStrategy/OnEventStrategy.php]: remove return

File: src/Plugin/InvokeStrategy/OnEventStrategy.php [hunk 32]
```diff
@@ -35,8 +36,10 @@ public function canInvoke($handler, $message)
     /**
      * @param mixed $handler
      * @param mixed $message
+     *
+     * @return void
```
Comment (basz) [inline src/Plugin/InvokeStrategy/OnEventStrategy.php]: remove return

File: src/Plugin/InvokeStrategy/OnEventStrategy.php [hunk 33]
```diff
@@ -45,11 +48,15 @@ public function invoke($handler, $message)
 
     /**
      * @param mixed $event
+     *
      * @return string
```
Comment (basz) [inline src/Plugin/InvokeStrategy/OnEventStrategy.php]: remove return

File: src/Plugin/Router/SingleHandlerRouter.php [hunk 34]
```diff
@@ -43,31 +43,22 @@ class SingleHandlerRouter implements MessageBusRouterPlugin, ActionEventListener
...
-    public function attach(ActionEventEmitter $events)
+    public function attach(ActionEventEmitter $events): void
     {
         $this->trackHandler($events->attachListener(MessageBus::EVENT_ROUTE, [$this, "onRouteMessage"]));
     }
 
-    /**
-     * @param string $messageName
-     * @return $this
-     * @throws Exception\RuntimeException
```
Comment (basz) [inline src/Plugin/Router/SingleHandlerRouter.php]: keep throws

File: src/CommandBus.php [hunk 35]
```diff
@@ -59,10 +55,12 @@ public function setActionEventEmitter(ActionEventEmitter $actionEventDispatcher)
 
     /**
      * @param mixed $command
-     * @throws CommandDispatchException
+     *
      * @return void
```
Comment (basz) [inline src/CommandBus.php]: remove return

File: src/Container/AbstractBusFactory.php [hunk 36]
```diff
@@ -64,57 +62,37 @@
...
             'enable_handler_location' => true,
             'message_factory' => MessageFactory::class,
         ];
     }
 
-    /**
-     * Create service.
-     *
-     * @param ContainerInterface $container
-     * @throws RuntimeException
```
Comment (basz) [inline src/Container/AbstractBusFactory.php]: keep throws phpdoc

File: src/MessageBus.php [hunk 37]
```diff
@@ -49,22 +51,17 @@
 
     /**
      * @param mixed $message
-     * @return mixed|void depends on the bus type
+     *
+     * @return \React\Promise\Promise|void depends on the bus type
```
Comment (basz) [inline src/MessageBus.php]: import namespace

File: src/MessageBus.php [hunk 38]
```diff
@@ -97,12 +94,7 @@ protected function initialize($message, ActionEvent $actionEvent)
         }
     }
 
-    /**
-     * @param ActionEvent $actionEvent
-     * @param \Exception $ex
-     * @throws Exception\MessageDispatchException
```
Comment (basz) [inline src/MessageBus.php]: keep throw, and import namespace

File: src/MessageBus.php [hunk 39]
```diff
@@ -117,11 +109,7 @@ protected function handleException(ActionEvent $actionEvent, \Exception $ex)
         }
     }
 
-    /**
-     * @param ActionEvent $actionEvent
-     * @throws Exception\RuntimeException
```
Comment (basz) [inline src/MessageBus.php]: keep throws

**Pull request merged**